diff --git a/CMakeLists.txt b/CMakeLists.txt
index a706f439..b3043e19 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -696,7 +696,11 @@ endif()
 string(TOLOWER "${DIST_ARCH}" DIST_ARCH)
 string(REPLACE "amd64" "x86_64" DIST_ARCH "${DIST_ARCH}")
 if (NOT DIST_DIR)
-    set(DIST_DIR "${DIST_ARCH}")
+    if(PLATFORM_NAME)
+        set(DIST_DIR "${DIST_ARCH}-${PLATFORM_NAME}")
+    else()
+        set(DIST_DIR "${DIST_ARCH}")
+    endif()
 endif()
 
 # ==================================================================================================
diff --git a/build.sh b/build.sh
index ecabc0fb..da9a40f2 100755
--- a/build.sh
+++ b/build.sh
@@ -679,9 +679,9 @@ function build_ios_target {
     local platform=$3
 
     echo "Building iOS ${lc_target} (${arch}) for ${platform}..."
-    mkdir -p "out/cmake-ios-${lc_target}-${arch}"
+    mkdir -p "out/cmake-ios-${lc_target}-${arch}-${platform}"
 
-    pushd "out/cmake-ios-${lc_target}-${arch}" > /dev/null
+    pushd "out/cmake-ios-${lc_target}-${arch}-${platform}" > /dev/null
 
     if [[ ! -d "CMakeFiles" ]] || [[ "${ISSUE_CMAKE_ALWAYS}" == "true" ]]; then
         cmake \
@@ -699,14 +699,14 @@ function build_ios_target {
             ${MATOPT_OPTION} \
             ${STEREOSCOPIC_OPTION} \
             ../..
-        ln -sf "out/cmake-ios-${lc_target}-${arch}/compile_commands.json" \
+        ln -sf "out/cmake-ios-${lc_target}-${arch}-${platform}/compile_commands.json" \
            ../../compile_commands.json
     fi
 
     ${BUILD_COMMAND}
 
     if [[ "${INSTALL_COMMAND}" ]]; then
-        echo "Installing ${lc_target} in out/${lc_target}/filament..."
+        echo "Installing ${lc_target}-${arch}-${platform} in out/${lc_target}/filament..."
         ${BUILD_COMMAND} ${INSTALL_COMMAND}
     fi
 
@@ -741,16 +741,26 @@ function build_ios {
     if [[ "${ISSUE_DEBUG_BUILD}" == "true" ]]; then
         build_ios_target "Debug" "arm64" "iphoneos"
         if [[ "${IOS_BUILD_SIMULATOR}" == "true" ]]; then
+            build_ios_target "Debug" "arm64" "iphonesimulator"
             build_ios_target "Debug" "x86_64" "iphonesimulator"
         fi
 
         if [[ "${BUILD_UNIVERSAL_LIBRARIES}" == "true" ]]; then
             build/ios/create-universal-libs.sh \
-                -o out/ios-debug/filament/lib/universal \
-                out/ios-debug/filament/lib/arm64 \
-                out/ios-debug/filament/lib/x86_64
-            rm -rf out/ios-debug/filament/lib/arm64
-            rm -rf out/ios-debug/filament/lib/x86_64
+                -o ./out/ios-debug/filament/lib/universal \
+                ./out/ios-debug/filament/lib/arm64-iphonesimulator \
+                ./out/ios-debug/filament/lib/x86_64-iphonesimulator
+
+            rm -rf out/ios-debug/filament/lib/x86_64-iphonesimulator
+            rm -rf out/ios-debug/filament/lib/arm64-iphonesimulator
+
+            build/ios/create-xc-frameworks.sh \
+                -o out/ios-debug/filament/lib \
+                out/ios-debug/filament/lib/arm64-iphoneos \
+                out/ios-debug/filament/lib/universal
+
+            rm -rf out/ios-debug/filament/lib/universal
+            rm -rf out/ios-debug/filament/lib/arm64-iphoneos
         fi
 
         archive_ios "Debug"
@@ -758,17 +768,28 @@ function build_ios {
 
     if [[ "${ISSUE_RELEASE_BUILD}" == "true" ]]; then
         build_ios_target "Release" "arm64" "iphoneos"
+
         if [[ "${IOS_BUILD_SIMULATOR}" == "true" ]]; then
+            build_ios_target "Release" "arm64" "iphonesimulator"
             build_ios_target "Release" "x86_64" "iphonesimulator"
         fi
 
         if [[ "${BUILD_UNIVERSAL_LIBRARIES}" == "true" ]]; then
             build/ios/create-universal-libs.sh \
-                -o out/ios-release/filament/lib/universal \
-                out/ios-release/filament/lib/arm64 \
-                out/ios-release/filament/lib/x86_64
-            rm -rf out/ios-release/filament/lib/arm64
-            rm -rf out/ios-release/filament/lib/x86_64
+                -o ./out/ios-release/filament/lib/universal \
+                ./out/ios-release/filament/lib/arm64-iphonesimulator \
+                ./out/ios-release/filament/lib/x86_64-iphonesimulator
+
+            rm -rf out/ios-release/filament/lib/x86_64-iphonesimulator
+            rm -rf out/ios-release/filament/lib/arm64-iphonesimulator
+
+            build/ios/create-xc-frameworks.sh \
+                -o out/ios-release/filament/lib \
+                out/ios-release/filament/lib/arm64-iphoneos \
+                out/ios-release/filament/lib/universal
+
+            rm -rf out/ios-release/filament/lib/universal
+            rm -rf out/ios-release/filament/lib/arm64-iphoneos
         fi
 
         archive_ios "Release"
diff --git a/third_party/clang/iOS.cmake b/third_party/clang/iOS.cmake
index a935c08f..ea491860 100644
--- a/third_party/clang/iOS.cmake
+++ b/third_party/clang/iOS.cmake
@@ -29,8 +29,17 @@ SET(CMAKE_CXX_COMPILER_WORKS True)
 SET(CMAKE_C_COMPILER_WORKS True)
 SET(DARWIN_TARGET_OS_NAME ios)
 
-SET(PLATFORM_NAME "iphoneos" CACHE STRING "iOS platform to build for")
-SET(PLATFORM_FLAG_NAME ios)
+IF(NOT DEFINED PLATFORM_NAME)
+    SET(PLATFORM_NAME "iphoneos" CACHE STRING "iOS platform to build for")
+ENDIF()
+
+IF(PLATFORM_NAME STREQUAL "iphonesimulator")
+    SET(PLATFORM_FLAG_NAME ios-simulator)
+    MESSAGE(STATUS "Building for iOS Simulator - PLATFORM_FLAG_NAME=${PLATFORM_FLAG_NAME}")
+ELSE()
+    SET(PLATFORM_FLAG_NAME ios)
+    MESSAGE(STATUS "Building for iOS Device - PLATFORM_FLAG_NAME=${PLATFORM_FLAG_NAME}")
+ENDIF()
 
 IF("$ENV{RC_APPLETV}" STREQUAL "YES")
   MESSAGE(STATUS "Building for tvos")
diff --git a/build/ios/create-xc-frameworks.sh b/build/ios/create-xc-frameworks.sh
new file mode 100755
index 00000000..0314d09b
--- /dev/null
+++ b/build/ios/create-xc-frameworks.sh
@@ -0,0 +1,92 @@
+#!/usr/bin/env bash
+
+set -e
+
+function print_help {
+    local SELF_NAME
+    SELF_NAME=$(basename "$0")
+    echo "$SELF_NAME. Create XCFrameworks from device and simulator libraries."
+    echo ""
+    echo "Usage:"
+    echo "    $SELF_NAME [options] <device_path> <simulator_path>"
+    echo ""
+    echo "Options:"
+    echo "    -h"
+    echo "       Print this help message."
+    echo "    -o"
+    echo "       Output directory to store the XCFrameworks."
+    echo ""
+    echo "Example:"
+    echo "    $SELF_NAME -o out/ios-release/filament/lib \\"
+    echo "        out/ios-release/filament/lib/arm64-iphoneos \\"
+    echo "        out/ios-release/filament/lib/universal"
+    echo ""
+}
+
+OUTPUT_DIR=""
+while getopts "ho:" opt; do
+    case ${opt} in
+        h)
+            print_help
+            exit 1
+            ;;
+        o)
+            OUTPUT_DIR="${OPTARG}"
+            ;;
+        *)
+            print_help
+            exit 1
+            ;;
+    esac
+done
+
+shift $((OPTIND - 1))
+
+DEVICE_PATH="$1"
+SIMULATOR_PATH="$2"
+
+if [[ ! "${DEVICE_PATH}" ]] || [[ ! "${SIMULATOR_PATH}" ]]; then
+    echo "Both device and simulator paths required."
+    print_help
+    exit 1
+fi
+
+if [[ ! "${OUTPUT_DIR}" ]]; then
+    echo "Output directory required."
+    print_help
+    exit 1
+fi
+
+# Create the output directory, if it doesn't exist already.
+mkdir -p "${OUTPUT_DIR}"
+
+# Loop through each library in the device path and create an XCFramework
+for FILE in "${DEVICE_PATH}"/*.a; do
+    [ -f "${FILE}" ] || continue
+
+    # The static library file name, like "libfilament.a"
+    LIBRARY_NAME="${FILE##*/}"
+
+    # Framework name without extension, like "libfilament"
+    FRAMEWORK_NAME="${LIBRARY_NAME%.a}"
+
+    echo "Creating XCFramework for library: ${LIBRARY_NAME}"
+
+    SIMULATOR_FILE="${SIMULATOR_PATH}/${LIBRARY_NAME}"
+    if [[ ! -f "${SIMULATOR_FILE}" ]]; then
+        echo "Warning: ${SIMULATOR_FILE} does not exist. Skipping."
+        continue
+    fi
+
+    # Remove existing XCFramework if it exists
+    XCFRAMEWORK_PATH="${OUTPUT_DIR}/${FRAMEWORK_NAME}.xcframework"
+    if [[ -d "${XCFRAMEWORK_PATH}" ]]; then
+        rm -rf "${XCFRAMEWORK_PATH}"
+    fi
+
+    # Create XCFramework
+    xcodebuild -create-xcframework \
+        -library "${FILE}" \
+        -library "${SIMULATOR_FILE}" \
+        -output "${XCFRAMEWORK_PATH}"
+done
